### PR TITLE
geohashesInbox(lon_min, lat_min, lon_max, lat_max, precision) function

### PR DESCRIPTION
**Pull request by Enmk**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

For changelog. Remove if this is non-significant change.

Category (leave one):
- New Feature

Short description (up to few sentences):
Function geohashesInBox() creates array of geohash-encoded boxes that cover given area.
...

Detailed description (optional):
- Function geohashesInBox() creates array of geohash-encoded boxes that cover given area. Resulting array of geohashes covers provided boundaries and more. Due to potential of exploding with multitude of values, this function has hardcoded limit of 10 million on produced array size.
- Tests and docs
...

### CODE REVIEW

File: dbms/src/Functions/geohashCoverBox.cpp [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+    void executeImpl(Block & block, const ColumnNumbers & arguments, size_t result, size_t /*input_rows_count*/) override
+    {
+        const IColumn * lon_min = block.getByPosition(arguments[0]).column.get();
+        const IColumn * lat_min = block.getByPosition(arguments[1]).column.get();
+        const IColumn * lon_max = block.getByPosition(arguments[2]).column.get();
+        const IColumn * lat_max = block.getByPosition(arguments[3]).column.get();
+        const IColumn * prec =    block.getByPosition(arguments[4]).column.get();
+        ColumnPtr & res = block.getByPosition(result).column;
+
+        if (tryExecute<Float32, Float32, Float32, Float32, UInt8>(lon_min, lat_min, lon_max, lat_max, prec, res) ||
```
Comment (akuzm) [inline dbms/src/Functions/geohashCoverBox.cpp]: If these four template parameters are always going to be the same, one is enough.
Also we can just determine the type from the first argument and call the right function. We've already checked that types are the same in getReturnTypeImpl.

**Comment by alexey-milovidov** [comment]
Is it officially named "cover box"?
Maybe "bounding box" is a better name?

**Comment by alexey-milovidov** [comment]
If it returns multiple boxes, maybe `*Boxes` is a better name?

File: dbms/src/Functions/geohashCoverBox.cpp [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+        UInt32 pos = 0;
+        for (size_t row = 0; row < total_rows; ++row)
+        {
+            const Float64 lon_min_value = lon_min->getElement(row);
+            const Float64 lat_min_value = lat_min->getElement(row);
+            const Float64 lon_max_value = lon_max->getElement(row);
+            const Float64 lat_max_value = lat_max->getElement(row);
+            const PrecisionType precision_value = GeoUtils::geohashPrecision(precision->getElement(row));
+
+            const auto positions = GeoUtils::geohashCoverBox(lon_min_value, lat_min_value,
```
Comment (akuzm) [inline dbms/src/Functions/geohashCoverBox.cpp]: Can we skip the intermediate vector here, and construct the resulting array right away?

File: dbms/src/Functions/geohashCoverBox.cpp (hunk 1 above)
Comment (Enmk) [reply dbms/src/Functions/geohashCoverBox.cpp]: well, I don't know how convenient having all arguments of the same type to the user, so I've left these template arguments as a backup plan. On the other hand I don't want to blow up code size with N specializations of that template.

**Comment by Enmk** [comment]
Regarding name: I personally don't think that `boundingBox` is a good choice, since that has a well-defined meaning. `boundingBoxes` is kind of confusing in that regard.

Given the rule that the more specific functions is, longer the name, maybe we shall rename that to `geohashCoverAreaWithBoxes()` ?

File: dbms/src/Functions/geohashCoverBox.cpp (hunk 2 above)
Comment (Enmk) [reply dbms/src/Functions/geohashCoverBox.cpp]: Well, I think it is possible. Let me try

**Comment by alexey-milovidov** [comment]
> Regarding name: I personally don't think that `boundingBox` is a good choice, since that has a well-defined meaning. `boundingBoxes` is kind of confusing in that regard.
> 
> Given the rule that the more specific functions is, longer the name, maybe we shall rename that to `geohashCoverAreaWithBoxes()` ?

Ok.

File: dbms/src/Functions/geohashCoverBox.cpp (hunk 2 above)
Comment (Enmk) [reply dbms/src/Functions/geohashCoverBox.cpp]: Done

File: dbms/src/Functions/geohashCoverBox.cpp (hunk 1 above)
Comment (Enmk) [reply dbms/src/Functions/geohashCoverBox.cpp]: Done

**Comment by alex-zaitsev** [comment]
geohashCoverAreaWithBoxes() seems confusing to me in 'WithBoxes' part. It assumes there are multiple boxes and we do something with them. But there is single 'box argument'. Bounding Box is a commonly used term. Probably we could name it geohashArrayFromBoundingBox() if you want a long descriptive name?

File: dbms/src/Functions/geohashCoverAreaWithBoxes.cpp [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+                    ErrorCodes::ARGUMENT_OUT_OF_BOUND};
+            }
+
+            res_strings_offsets.reserve(res_strings_offsets.size() + estimated_array_size);
+            res_strings_chars.resize(res_strings_chars.size() + estimated_array_size * (precision_value + 1));
+            const auto starting_offset = res_strings_offsets.empty() ? 0 : res_strings_offsets.back();
+            char * out = reinterpret_cast<char *>(res_strings_chars.data() + starting_offset);
+
+            const auto actual_array_size = GeoUtils::geohashCoverAreaWithBoxes(
+                        lon_min_value, lat_min_value, lon_max_value, lat_max_value, precision_value, out);
```
Comment (akuzm) [inline dbms/src/Functions/geohashCoverAreaWithBoxes.cpp]: Can you just throw in geohashCover.... if you have too many boxes? You won't need all the other functions then. Or maybe even code all this in place. Having three embedded loops is slightly inconvenient, but they are very simple so it's probably ok.

File: dbms/src/Functions/geohashCoverAreaWithBoxes.cpp [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+        const IColumn * lon_min = block.getByPosition(arguments[0]).column.get();
+        const IColumn * lat_min = block.getByPosition(arguments[1]).column.get();
+        const IColumn * lon_max = block.getByPosition(arguments[2]).column.get();
+        const IColumn * lat_max = block.getByPosition(arguments[3]).column.get();
+        const IColumn * prec =    block.getByPosition(arguments[4]).column.get();
+        ColumnPtr & res = block.getByPosition(result).column;
+
+        if (tryExecute<Float32, UInt8>(lon_min, lat_min, lon_max, lat_max, prec, res) ||
+            tryExecute<Float64, UInt8>(lon_min, lat_min, lon_max, lat_max, prec, res))
+            return;
```
Comment (akuzm) [inline dbms/src/Functions/geohashCoverAreaWithBoxes.cpp]: Looks better with two template parameters. 
Now, if the second one is the same, we probably don't need it as well.

Another thing it that this checking logic just looks strange. First you check the argument types in getReturnTypeImpl.
Then, you try to execute it one way, if it fails -- another way, then finally error out again. Why is this necessary? You know the column type, so you can just call the right function. If tryExecute encounters a wrong type, it is a logical error. So basically this entire function should be replaced with `if (column is float32) execute<Float32>() else execute<Float64>()`.

File: dbms/src/DataTypes/IDataType.h [hunk 5]
```diff
@@ -603,6 +603,12 @@ inline bool isStringOrFixedString(const T & data_type)
     return WhichDataType(data_type).isStringOrFixedString();
 }
 
+template <typename T, typename U>
+inline bool isSame(const T & data_type_left, const U & data_type_right)
```
Comment (akuzm) [inline dbms/src/DataTypes/IDataType.h]: There is IDataType::equals().

File: dbms/src/Functions/geohashCoverAreaWithBoxes.cpp (hunk 3 above)
Comment (Enmk) [reply dbms/src/Functions/geohashCoverAreaWithBoxes.cpp]: I don't see why, most of the code here is regarding forcing data into Array of String-s, so removing a tiny bit that throws an exception wouldn't bring much simplicity.

* `geohashPrecision` is required to properly estimate string storage and properly compute string offsets.
* `geohashEstimateCoverAreaWithBoxesItems` is required to estimate string storage too.

Comment (akuzm) [reply dbms/src/Functions/geohashCoverAreaWithBoxes.cpp]: I mean the `prologue` function. Apparently it calculates the number of small boxes -- couldn't tell from the name. It returns a three field tuple -- I already forgot which field is which. You call it twice -- first, through a single-line wrapper (does the wrapper really add anything?) to get the size of resulting array. Then it is called again -- now by the function that creates the boxes. So the calculation is done twice (well ok it's fast), and two places in code have the assumption that they do the same calculation (again, doesn't matter here but in general that's ugly). If you have to split this in several functions, at least make a struct for the state of the algorithm, that has properly named fields, and calculate them once. You can put input parameters there as well.

**Comment by Enmk** [comment]
> geohashCoverAreaWithBoxes() seems confusing to me in 'WithBoxes' part. It assumes there are multiple boxes and we do something with them. But there is single 'box argument'. Bounding Box is a commonly used term. Probably we could name it geohashArrayFromBoundingBox() if you want a long descriptive name?

As agreed on slack, we'd use the `geohashesInBox` name.

File: dbms/src/Functions/geohashCoverAreaWithBoxes.cpp (hunk 3 above)
Comment (Enmk) [reply dbms/src/Functions/geohashCoverAreaWithBoxes.cpp]: Saved one computation, made named struct, reduced number of geohash-related functions in `GeoUtils`

File: dbms/src/Functions/geohashCoverAreaWithBoxes.cpp (hunk 4 above)
Comment (Enmk) [reply dbms/src/Functions/geohashCoverAreaWithBoxes.cpp]: It looks like both approaches (existing and proposed) already exist in codebase, so it is not quite clear which one is more preferable. Take a look at:
* FunctionNumericPredicate
* FunctionBitTestMany
* FunctionToStringCutToZero
* FunctionHex

And many others, for this we won't win anything in terms of simplicity nor amount of generated code, and win only a little in terms of performance (constant number of calls to `checkAndGetColumn<T>`) here is an example at call site:

proposed
```
        auto * precision = checkAndGetColumn<ColumnVector<PrecisionType>>(precision_column);

        if (const auto * lon_min = checkAndGetColumn<ColumnVector<Float64>>(lon_min_column))
        {
	        const auto * lat_min = checkAndGetColumn<ColumnVector<Float64>>(lat_min_column);
	        const auto * lon_max = checkAndGetColumn<ColumnVector<Float64>>(lon_max_column);
	        const auto * lat_max = checkAndGetColumn<ColumnVector<Float64>>(lat_max_column);
	        execute(lon_min, lat_min, lon_max, lon_max, precision, result);
        }
        else if (const auto * lon_min = checkAndGetColumn<ColumnVector<Float32>>(lon_min_column))
        {
	        const auto * lat_min = checkAndGetColumn<ColumnVector<Float32>>(lat_min_column);
	        const auto * lon_max = checkAndGetColumn<ColumnVector<Float32>>(lon_max_column);
	        const auto * lat_max = checkAndGetColumn<ColumnVector<Float32>>(lat_max_column);
	        execute(lon_min, lat_min, lon_max, lon_max, precision, result);
        }
```

vs

current
```
        if (tryExecute<Float32, UInt8>(lon_min, lat_min, lon_max, lat_max, prec, res) ||
            tryExecute<Float64, UInt8>(lon_min, lat_min, lon_max, lat_max, prec, res))
            return;
```

File: dbms/src/DataTypes/IDataType.h (hunk 5 above)
Comment (Enmk) [reply dbms/src/DataTypes/IDataType.h]: fixed

File: dbms/src/Functions/geohashCoverAreaWithBoxes.cpp (hunk 4 above)
Comment (akuzm) [reply dbms/src/Functions/geohashCoverAreaWithBoxes.cpp]: Ok, both approaches exist, so how do we choose one or another? If checking the arguments twice is not necessary, we shouldn't do it. Or is it necessary, i.e. is there a case when we check the argument types in `getReturnTypeImpl`, but still get wrong types in `executeImpl`? What is this case, should we add it to the tests and mention it in the comments? Without any comments, the reason for double checking is not obvious. If the double checking is really required, this is a problem of Function interface and not of this PR, but let's make this clear.

To make the code shorter, we can write: 
`if (lat_min_column.isFloat32()) execute<Float32>(...) else execute<Float64>(...).` The execute() function should throw logical error if it encounters a column of a wrong type.

**Comment by akuzm** [comment]
I noticed that we get an empty array for point (45, 45, 45, 45). Is this expected? 
```
┌────n─┬─geohashesInBox(n, n, n, n, 1)─┐
│ 44.9 │ ['s']                         │
│   45 │ []                            │
│ 45.1 │ ['v']                         │
└──────┴───────────────────────────────┘
```

**Comment by Enmk** [comment]
> I noticed that we get an empty array for point (45, 45, 45, 45). Is this expected?
> 
> ```
> ┌────n─┬─geohashesInBox(n, n, n, n, 1)─┐
> │ 44.9 │ ['s']                         │
> │   45 │ []                            │
> │ 45.1 │ ['v']                         │
> └──────┴───────────────────────────────┘
> ```

You just hit a jackpot, a very good find! Nope, this is not expected, this is a very edge case of a empty range on a geohash grid line. But it is fixed now.

File: dbms/src/Functions/geohashCoverAreaWithBoxes.cpp (hunk 4 above)
Comment (Enmk) [reply dbms/src/Functions/geohashCoverAreaWithBoxes.cpp]: Fixed

File: dbms/src/Functions/GeoUtils.cpp [hunk 6]
```diff
@@ -60,11 +66,10 @@ inline UInt8 singleCoordBitsPrecision(UInt8 precision, CoordType type)
 inline Encoded encodeCoordinate(Float64 coord, Float64 min, Float64 max, UInt8 bits)
 {
     Encoded result;
-    result.fill(0);
```
Comment (akuzm) [inline dbms/src/Functions/GeoUtils.cpp]: std::array doesn't initialize values by default, so the higher bits are going to be uninitialized. Probably they are not used, but I'd prefer to not have uninitialized values. Is this initialization becoming a performance bottleneck? This encoding looks suboptimal, it basically does the shift and division bit-by-bit. I think we can optimize it if needed. But this should be a different PR.

I'll remove this change myself if you don't mind, so you don't have to update the PR.

**Pull request merged**